File: 110-balanced-binary-tree/sol-2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
  var [balanced, depth] = validate(root)
  return balanced
}

function validate(root) {
  if (!root) {
    return [true, 0]
  }

  var [leftBalanced, leftDepth] = validate(root.left)
  if (!leftBalanced) {
    return [false, 0]
  }

  var [rightBalaced, rightDepth] = validate(root.right)
  if (!rightBalaced) {
    return [false, 0]
  }

  return [Math.abs(leftDepth - rightDepth) <= 1, Math.max(leftDepth, rightDepth) + 1]
}

File: 101-200/110-balanced-binary-tree/sol-2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):

    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        balanced, _ = self.validate(root)
        return balanced

    def validate(self, root):
        if not root:
            return True, 0
        balanced, leftDepth = self.validate(root.left)
        if not balanced:
            return False, 0
        balanced, rightDepth = self.validate(root.right)
        if not balanced:
            return False, 0
        return abs(leftDepth - rightDepth) <= 1, max(leftDepth, rightDepth) + 1
